### ISSUE TITLE
std::source_location line number wrong when used as default argument value

### ISSUE THREAD

**Issue by IGR2014** [llvm/llvm-project#56379]
Consider the following code snipet ([CompilerExplorer link](https://godbolt.org/z/3h3n5Moj6)):
```c++
#include<source_location>
#include<iostream>

void test_source_location(std::source_location src = std::source_location::current()) {
    std::cout << "called at " << src.line() << "\n";
}

int main() {
    test_source_location();
}
```

The output results for Clang, GCC and MSVC (left-to-right) are:
![image](https://user-images.githubusercontent.com/10415461/177286976-dd69339c-bd49-4f2a-9349-ae3bd1bdf2fd.png)

As you can see - both GCC and MSVC reports function INVOCATION line number, while Clang report function DECLARATION-one

**Comment by ilya-biryukov** [comment]
I will take look at this

**Comment by aasimon** [comment]
I would love to help out on this task if I can. So let me know if there is anything I could do to help.

**Comment by ilya-biryukov** [comment]
Progress update from me.
I have a half-working fix ([D129488](https://reviews.llvm.org/D129488)), but it breaks consteval functions (see comments).
@jacobsa was about to pick it up from there.

**Comment by ilya-biryukov** [comment]
@aasimon we'll let you know if we don't make progress here, so you can pick it up.
Unless you feel really strongly about taking this over.

**Comment by aasimon** [comment]
> @aasimon we'll let you know if we don't make progress here, so you can pick it up. Unless you feel really strongly about taking this over.

I hardly think I am qualified for taking over the task ;-) I was more referring to helping out with testing and such.

**Comment by jacobsa** [comment]
Just for the record, the standard does agree that the current behavior is incorrect. From [[support.srcloc.cons]/2](https://timsong-cpp.github.io/cppwp/n4868/support.srcloc#cons-2) in N4868:

> Any call to `current` that appears as a default argument ([dcl.fct.default]), or as a subexpression thereof, should correspond to the location of the invocation of the function that uses the default argument ([expr.call]).

**Comment by ChuanqiXu9** [comment]
> Just for the record, the standard does agree that the current behavior is incorrect. From [[support.srcloc.cons]/2](https://timsong-cpp.github.io/cppwp/n4868/support.srcloc#cons-2) in N4868:
> 
> > Any call to `current` that appears as a default argument ([dcl.fct.default]), or as a subexpression thereof, should correspond to the location of the invocation of the function that uses the default argument ([expr.call]).

Currently the wording is:
> Any call to current that appears as a default member initializer ([[class.mem]](https://timsong-cpp.github.io/cppwp/n4868/class.mem)), or as a subexpression thereof, should correspond to the location of the constructor definition or aggregate initialization that uses the default member initializer[.](https://timsong-cpp.github.io/cppwp/n4868/support.srcloc#cons-2.sentence-1)

But the example is not default member initializer.

**Comment by jacobsa** [comment]
I'm not sure what you mean. The sentence I quoted comes after the one you quoted, right?

**Comment by ChuanqiXu9** [comment]
> I'm not sure what you mean. The sentence I quoted comes after the one you quoted, right?

Yes, my bad for oversee.

**Comment by cor3ntin** [comment]
I'm currently working on implementing [CWG2631](https://cplusplus.github.io/CWG/issues/2631.html) to resolve this issue.

**Comment by cor3ntin** [comment]
I'm doing multiple things, all of which are probably necessary / useful

- Implement CWG2631 by delaying immediate functions evaluation, and rebuilding expressions when they are used. I avoid rewriting when there is no immediate call. I suspect we will struggle properly rewriting lambdas, blocks...
- Treat `__builtin_source_location*` as immediate "calls"
- Instead of having overwriting scopes of default arguments , we need to keep track of the outermost default argument/init list
- Then we should be able to remove the codegen code path for source location

@AaronBallman

**Comment by jacobsa** [comment]
Thanks for looking. I should have updated above: I kind of gave up on this, because it's a little more involved than expected and I don't have a lot of experience within clang proper.

**Comment by ChuanqiXu9** [comment]
@cor3ntin hi, although I didn't look the problems in detail, I believe the proper fix will address the issue too: https://github.com/llvm/llvm-project/issues/57459

Hope this helps.

**Comment by cor3ntin** [comment]
@ChuanqiXu9 There is more work to be done for modules to support that, maybe you could take a look once my PR is merged?
I'm not familiar with how export / reachable works in clang

**Comment by ChuanqiXu9** [comment]
> maybe you could take a look once my PR is merged?

Sure. I looked at your WIP patches and it seems like more complex than I imaged. I imaged we'll transform the calls with default arguments to the general style. For example, https://godbolt.org/z/vx4x5Er8h I imaged with the imaged "proper" fix, the dumped AST would be:

```
BinaryOperator 0x561a91dfd9b0 <col:12, col:27> 'int' '+'
        |-CallExpr 0x561a91ddd8c0 <col:12, col:16> 'int'
        | |-ImplicitCastExpr 0x561a91ddd8a8 <col:12> 'int (*)(B)' <FunctionToPointerDecay>
        | | `-DeclRefExpr 0x561a91ddd830 <col:12> 'int (B)' lvalue Function 0x561a91ddd5b8 'foo' 'int (B)'
        | `-CXXTemporaryObjectExpr 0x561a91ddd8e8 <<invalid sloc>> 'B':'B'
        `-CallExpr 0x561a91dfd988 <col:20, col:27> 'int'
          |-ImplicitCastExpr 0x561a91dfd970 <col:20> 'int (*)(B)' <FunctionToPointerDecay>
          | `-DeclRefExpr 0x561a91dfd950 <col:20> 'int (B)' lvalue Function 0x561a91ddd5b8 'foo' 'int (B)'
          `-CXXTemporaryObjectExpr 0x561a91dfd920 <col:24, col:26> 'B':'B' 'void () noexcept' zeroing
```

Then we'll get things addressed correctly. But I may took things too easily. There may be more works on the module's semantics' side.

**Comment by cor3ntin** [comment]
This will be fixed by https://reviews.llvm.org/D136554

**Comment by cor3ntin** [comment]
This is now fixed on main.

**Comment by aasimon** [comment]
Any idea which clang version the fix will end up being released in?

**Comment by AaronBallman** [comment]
> Any idea which clang version the fix will end up being released in?

It's expected to be released in Clang 16.

**Comment by cor3ntin** [comment]
Note that this was reverted temporarily as we found bugs in the fix, but we are still targeting Clang 16

**Comment by avikivity** [comment]
Please backport it to clang 15.0.x

**Comment by avikivity** [comment]
And, if the fix has not landed again, I suggest to re-open the bug.

**Comment by cor3ntin** [comment]
Alas, it's a pretty involved changed, both in the compiler and in the language, so it's not a great candidate for backporting.
I'm still working through a few issues but I'm hoping to be able to re-reland that shortly.

**Comment by jyknight** [comment]
@cor3ntin re-landed it ca619613801233ef2def8c3cc7d311d5ed0033cb a few days ago, looks like it's stick this time, so:
Fixed for Clang 16. 🎉
